### PR TITLE
Patched digest-0.0.1.2

### PR DESCRIPTION
**Pull request by nightscape**
First time contributor here :smiley:
I'm trying to make [Idris](https://github.com/idris-lang/Idris-dev) compile with Eta.
It's probably a long shot, but I think I'll be having fun along the way, so this is the first step.
I'm a relative Haskell noob, so I might have made stupid mistakes :stuck_out_tongue_winking_eye:
If so, I'm happy for any pointers for improvement!

### CODE REVIEW

**Comment by puffnfresh** [comment]
Woo! Let me know if you need extra help with getting Idris to run!

File: patches/digest-0.0.1.2.patch [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
++import eta.runtime.io.MemoryManager;
++
++/* NOTE: Hard limit on ByteBuffer API: 2^32 - 1 because of reliance on
++         signed integer type everywhere. */
++
++public class Utils {
++
++    public static long calc_crc32(int seed, long address, int len) {
++        ByteBuffer buf = MemoryManager.getBoundedBuffer(address);
++        return Digest.calc_crc32(seed, buf);
```
Comment (puffnfresh) [inline patches/digest-0.0.1.2.patch]: For some reason this isn't found under 0.0.9

Comment (nightscape) [reply patches/digest-0.0.1.2.patch]: Ah, sorry. Copy & paste error. I added a fixup. If it works for you with that, I'll squash it.

**Comment by rahulmutt** [comment]
Awesome work @nightscape! Looking forward to seeing the Idris compiler compiled with Eta. I'll have to do a careful review of this to make sure you translated the unsigned operations (the trickiest bit in porting C APIs correctly). 

Have you done any tests by setting up a test project and seeing the outputs of GHC vs Eta?

**Comment by nightscape** [comment]
Yes, that unsigned part was what took me the longest time because I couldn't reuse `java.util.zip.CRC32` because of the lack of a way to update a CRC32.

There is a testing folder which contains two files called `trivial.expected` and `trivial.hs`.
I added a `executable` section to `digest.cabal`, ran `trivial.hs` and compared the output to `trivial.expected`. I just had to comment out the test with `\0` as that lead to a Java exception about the generated class containing an invalid String.
Apart from that the results were identical.

**Comment by nightscape** [comment]
And another fixup... I forgot to add the updated `.cabal` file as well...
Btw, why does one create both a patch that contains the changes to the `.cabal` file as well as the file itself?

**Comment by rahulmutt** [comment]
It's because of how `etlas` works - it needs just the cabal file for the purpose of extracting dependency information. It was easy to implement at the time but I think we're ready to make it smarter by applying the patch file to the individual cabal file as well and avoid the need for storing cabal files. I'll file an issue on this.

**Comment by rahulmutt** [comment]
Great work! Merging. Feel free to resend PRs with updated patches if you find any subtle bugs with this one.

**Comment by nightscape** [comment]
Great! I would have rebased the code to squash the fixups but that's just my exceeding passion for cleanliness :wink:

**Comment by rahulmutt** [comment]
Don't worry about it :) This repository isn't something that people are gonna look at the history of so I'm not too concerned with cleanliness here.

**Pull request merged**